src/msw/webview_edge.cpp

```diff
@@ -26,10 +26,103 @@
 #include "wx/msw/private/cotaskmemptr.h"
 #include "wx/msw/private/webview_edge.h"
 
-#include <wrl/event.h>
-#include <Objbase.h>
+#include <objbase.h>
 
+#ifdef __VISUALC__
+#include <wrl/event.h>
 using namespace Microsoft::WRL;
+#else
+#include <atomic>
+
+template <typename baseT, typename ...argTs>
+class CInvokable : public baseT
+{
+public:
+    CInvokable() : m_nRefCount(0) {}
+    virtual ~CInvokable() {}
+    // IUnknown methods
+    HRESULT STDMETHODCALLTYPE QueryInterface(REFIID WXUNUSED(riid), void** ppvObj) override
+    {
+        /**
+         * WebView2 Runtime apparently doesn't use this method, so it doesn't
+         * matter how we implement this. On the other hand, this method must be
+         * implemented to make this invokable type a concrete class instead of a
+         * abstract one.
+         */
+        wxLogWarning(_("WebView2 called unimplemented CInvokable::QueryInterface"));
```

Comment:
IMO this should be a `wxLogDebug` and the message shouldn't be translated because I don't see what an end user might possibly do about this.

Reply:
I am not completely sure `CInvokable::QueryInterface` will not be called by WebView2, which is why I used `wxLogWarning`. It'd better to change to `wxLogDebug` in 3.1.6 if nobody complains about the warning message when using 3.1.5 if this is going to be merged into 3.1.5.

Reply:
Then this should be `wxFAIL_MSG()`. Or maybe we should just implement it because it's a clear violation of COM rules to not implement QI.

Reply:
Why did you said "not implement QI" when these is a QI calling `wxLogWarning`? It doesn't violate COM rules, it tries to comply with COM rules instead. Could you please do some researches yourself before popping up your so many opinions again? No offensive, but it really is annoying. I actually deleted my pull requests, because I was tired of having to explain every single trivial and OBVIOUS thing to you. Glad to see other people still interested in making (GCC/Clang + WebView2) to happen.

Reply:
@MaartenBent the other user is obviously wrong as a quick web search suffices to show (QI must work for `IUnknown` at the very least, that's the first point of [Raymond Chen's post](https://devblogs.microsoft.com/oldnewthing/20040326-00/?p=40033) from back in 2004 that they didn't even bother to read), but I won't discuss with them here as it's clearly counter-productive and will let you decide what to do here (of course, you can also decide to let me decide, whatever you prefer). Sorry for all the drama.

Reply:
I put a breakpoint in `QueryInterface` and it does not seem to be called when using the webview sample. I thought it could be useful to log a message (to the developer) when this might happen in the future.

According to that post it should respond to `IID_IUnknown` and its own interface?  Maybe something like:
```c++
if (riid == __uuidof(baseT) || riid == IID_IUnknown) {
    *ppvObj = this;
    AddRef();
    return S_OK;
}
```

Reply:
Yes, this is what I meant by "just implement it", and I've just checked and we don't use any intermediate interfaces as bases, i.e. there are no `ICoreWebView2XXX` that inherit from anything else than `IUnknown`, so this should be correct for all of them.

Just a question: does `__uuidof()` work with gcc now? I didn't know this, but it's great news if it does.

Reply:
> Just a question: does __uuidof() work with gcc now? 

It compiles. I suppose it will work as well. It has been part of the mingw64 header for at least 7 years.